### ISSUE TITLE
EULA insert users full names, device serial, asset name, etc

### ISSUE THREAD

**Issue by dborko95** [grokability/snipe-it#8607]
**Server (please complete the following information):**
 - Snipe-IT Version 5.0.2
 - OS: [e.g. Ubuntu]
 - Web Server: [Apache]
 - PHP Version 7.4

**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is.

We cant include user info or device info into a user agreement contract

**Describe the solution you'd like**

It would be awesome to have a group of variables that you can drag and drop into sections of the eula that will be filled out automatically upon assigning an asset

**Describe alternatives you've considered**

No alternatives I can think of compared to snipe-it

**Additional context**

@SjamonDaal built a great implementation of it:

https://github.com/snipe/snipe-it/issues/5933

Its exactly what I am looking for and I think would benefit a lot of people! If I didnt issue the link correctly I'm sorry, Im new to github. 

I also really appreciate you fixing issue #8585. Such a quick turn around!

**Comment by snipe** [comment]
I feel like we've heard this request before, but I can't find the exact GH issue I'm looking for for that. It's definitely something on the roadmap, but it gets a little squirrelly with stuff like the global EULA (versus category based) since we send those out for all kinds of different item types, and I can easily see chaos ensuing when people try to use a global EULA (referencing things like asset_tag) and that gets sent out on a license which doesn't have those attributes. 

I definitely think it can be done, but we'd need to figure out a way to make that less user-hostile than it could be.

**Comment by dborko95** [comment]
Aha makes sense. Thank you so much for considering it! Happy (hopefully) Friday

**Comment by snipe** [comment]
It's not impossible, we just really have to think it through so we don't end up making things ugly for the users. Maybe you can only use those variables when it's a category-specific EULA (versus the global one), so the category type (asset vs accessory vs licenses) would limit what variables might be available in those EULAs. It's as much a UX issue as it is a technical one, and the UX issues tend to take longer to figure out.

**Comment by SjamonDaal** [comment]
> Maybe you can only use those variables when it's a category-specific EULA (versus the global one), so the category type (asset vs accessory vs licenses) would limit what variables might be available in those EULAs.

Yeah right @snipe, this is exactly what I've built and [asked your opinion about](https://github.com/snipe/snipe-it/issues/5933#issuecomment-432972944) years ago. I known it wasn't finished and could have used some UX Improvement but all logic worked.

**Comment by SjamonDaal** [comment]
@dborko95, thanks for restarting this and mentioning me.

**Comment by robbindiepeveen** [comment]
@SjamonDaal, still having the dev code laying around somewhere in a dusty corner?
Interested in those features as well.

Thanks.

**Comment by SjamonDaal** [comment]
> @SjamonDaal, still having the dev code laying around somewhere in a dusty corner?
> Interested in those features as well.
> 
> Thanks.

I think I do. I'll take a look and post it on my git if I found it and I'm able/allowed to publish this.

**Comment by rickvandersterren** [comment]
> > @SjamonDaal, still having the dev code laying around somewhere in a dusty corner?
> > Interested in those features as well.
> > Thanks.
> 
> I think I do. I'll take a look and post it on my git if I found it and I'm able/allowed to publish this.

Djamon, this is just what I'm looking for. Couldn't find the repo on your account. Do you have it somewhere?

**Comment by SjamonDaal** [comment]
Yep I did find fhe code but I only have the code and no git logging so it's searching for changes in the code. This isn't worth

> > > @SjamonDaal, still having the dev code laying around somewhere in a dusty corner?
> > > Interested in those features as well.
> > > Thanks.
> > 
> > 
> > I think I do. I'll take a look and post it on my git if I found it and I'm able/allowed to publish this.
> 
> Djamon, this is just what I'm looking for. Couldn't find the repo on your account. Do you have it somewhere?

Yep, I did find it but I don't have the .git folder because it was excluded by some old crappy backup technology. There is no change log for it. It will be a challenge to find all of my changes.

Still interested? Then I'll try to find some time to post it on my github.

**Comment by robbindiepeveen** [comment]
I'm still interested as well! Thanks @SjamonDaal

**Comment by Almer** [comment]
> > > @SjamonDaal, still having the dev code laying around somewhere in a dusty corner?
> > > Interested in those features as well.
> > > Thanks.
> > 
> > 
> > I think I do. I'll take a look and post it on my git if I found it and I'm able/allowed to publish this.
> 
> Djamon, this is just what I'm looking for. Couldn't find the repo on your account. Do you have it somewhere?

I'm

> ble/allowed to publish thi

Hi Djamon, i'm interested too. Is there something possible to share the code? 

Looking forward to hear from you!

**Comment by SjamonDaal** [comment]
I'll be working on this and I'll post it here when I'm done. (It's a private project for now.)
https://github.com/SjamonDaal/snipe-it-with-eula-normalization

I still need to remove some static information we added for the organisation I worked at.

**Comment by robbindiepeveen** [comment]
Hi @SjamonDaal, any update on the progress?

**Comment by csayre** [comment]
I would love to see this implemented as well. Right now we are using paper forms, and the organization that I work for says we can go to digital if we can get this implemented. I'd love to help if I can, with testing or documenting. Not really that good a coding, but I can make things work if needed.

**Comment by 2ndSysAdmin** [comment]
@SjamonDaal any updates? We are looking for something like that too. Would be awesome to have.

**Comment by SjamonDaal** [comment]
Hi all,

I just want to say that I appreciate all the support for this feature from all of you. The implementation I've build in 2018 is very outdated and I don't have the time to completely refactor it. Before publishing old code I truly have to look at it whenever I've got time to spare. I also don't have the .git folder anymore only the source code, so my git history is gone and this removes any trace of what's changed. I don't want to publish it like that.

Update: I mean, I don't even know what version I'm at to refactor my changes. Tons of work and time I don't have and want to spend on this project atm.

**Comment by csayre** [comment]
I would still love to see this.

**Comment by snipe** [comment]
The main issues I had with @SjamonDaal’s implementation was that 

1.the variable placeholders are too complicated
2. it doesn’t account for null values in those placeholders - not everyone fills out all available fields, and we would have to build in sane defaults or conditionals for that so the phrasing doesn’t look weird
3. if an admin only has one EULA that they use for the whole system (via the admin settings, versus the category-based ones), those placeholders won’t always make sense, since there isn’t a concept of “seats” for assets, and not every asset has a name. Whlle we could limit the placeholders based on category-specific EULAs, people will definitely try to use them in the system-wide one, their EULAs will look weird, and then they’ll blame us

At the end of the day, the EULA pdfs and emails already show a lot of the item information above the EULA itself, so them signing it shouldn’t require the embedded stuff. 

We’ll obviously keep noodling on this, but unfortunately one of the bigger challenges we have here is to come up with solutions that work for *everyone*, not just a small subset of users.
